File: abstractions/dotnet/src/authentication/BaseBearerTokenAuthenticationProvider.cs
// ------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
// ------------------------------------------------------------------------------

using System;
using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.Kiota.Abstractions.Authentication;
/// <summary>
///     Provides a base class for implementing <see cref="IAuthenticationProvider" /> for Bearer token scheme.
/// </summary>
public class BaseBearerTokenAuthenticationProvider : IAuthenticationProvider
{
    /// <summary>
    /// Creates a new instance of <see cref="BaseBearerTokenAuthenticationProvider"/>.
    /// </summary>
    /// <param name="accessTokenProvider">The <see cref="IAccessTokenProvider"/> to use for getting the access token.</param>
    public BaseBearerTokenAuthenticationProvider(IAccessTokenProvider accessTokenProvider)
    {
        AccessTokenProvider = accessTokenProvider ?? throw new ArgumentNullException(nameof(accessTokenProvider));
    }
    /// <summary>
    ///     Gets the <see cref="IAccessTokenProvider" /> to use for getting the access token.
    /// </summary>
    public IAccessTokenProvider AccessTokenProvider {get; private set;}
    private const string AuthorizationHeaderKey = "Authorization";

    /// <summary>
    /// Authenticates the <see cref="RequestInformation"/> instance
    /// </summary>
    /// <param name="request">The request to authenticate</param>
    /// <param name="cancellationToken">The cancellation token for the task</param>
    /// <returns></returns>
    public async Task AuthenticateRequestAsync(RequestInformation request, CancellationToken cancellationToken = default)
    {
        if(request == null) throw new ArgumentNullException(nameof(request));
        if(!request.Headers.ContainsKey(AuthorizationHeaderKey))
        {
            var token = await AccessTokenProvider.GetAuthorizationTokenAsync(request.URI, cancellationToken);
            if(string.IsNullOrEmpty(token))
                throw new InvalidOperationException("Could not get an authorization token");
            request.Headers.Add(AuthorizationHeaderKey, $"Bearer {token}");
        }
    }
}

File: abstractions/php/src/Authentication/BaseBearerTokenAuthenticationProvider.php
<?php

namespace Microsoft\Kiota\Abstractions\Authentication;

use Http\Promise\FulfilledPromise;
use Http\Promise\Promise;
use Microsoft\Kiota\Abstractions\RequestInformation;

/**
 * Class BaseBearerTokenAuthenticationProvider
 *
 * Provides a base class for implementing {@link AuthenticationProvider} for Bearer authentication scheme
 *
 * @package Microsoft\Kiota\Abstractions\Authentication
 * @copyright 2022 Microsoft Corporation
 * @license https://opensource.org/licenses/MIT MIT License
 * @link https://developer.microsoft.com/graph
 */
class BaseBearerTokenAuthenticationProvider implements AuthenticationProvider {

    /**
     * @var string $authorizationHeaderKey The Authorization header key
     */
    private static string $authorizationHeaderKey = "Authorization";

    /**
     * @var AccessTokenProvider {@link AccessTokenProvider}
     */
    private AccessTokenProvider $accessTokenProvider;

    /**
     * Creates a new instance
     * @param AccessTokenProvider $accessTokenProvider to use for getting the access token
     */
    public function __construct(AccessTokenProvider $accessTokenProvider)
    {
        $this->accessTokenProvider = $accessTokenProvider;
    }

    /**
     * Gets the {@link AccessTokenProvider} used for getting the access token
     *
     * @return AccessTokenProvider
     */
    public function getAccessTokenProvider(): AccessTokenProvider
    {
        return $this->accessTokenProvider;
    }

    /**
     * @param RequestInformation $request
     * @return Promise
     */
    public function authenticateRequest(RequestInformation $request): Promise {
        if (!array_key_exists(self::$authorizationHeaderKey, $request->headers)) {
            return $this->accessTokenProvider->getAuthorizationTokenAsync($request->getUri())
                        ->then(function ($token) use($request) {
                            if ($token) {
                                $request->headers[self::$authorizationHeaderKey] = "Bearer {$token}";
                            }
                            return null;
                        });
        }
        return new FulfilledPromise(null);
    }
}
